### ISSUE TITLE
Cache file not created on certain files.

### ISSUE THREAD

**Issue by kdonbekci** [tmbdev/webdataset#53]
Caching consistently fails on certain `.tar` files when iterating over the dataset. The constructed cachefile remains in the temporary filepath. Upon sniff test of the problematic archives, there seems to be no apparent problem such as missing a file for a key or corrupted files. I manually iterated over the tars, extracted all the keys and compared them to the wds iteration. 

![image](https://user-images.githubusercontent.com/27675252/113335106-5c5ba300-92d9-11eb-9bf8-479ec2e824f3.png)

I've also created a Jupyter Colab here, demonstrating the issue.
https://colab.research.google.com/drive/1WNx5gqjcPpk-WLZsEu7XJmNWs9JRWpI9?usp=sharing

**Comment by tmbdev** [comment]
Shards only get cached after you have completely iterated through them. If you stop iterating on a shard, it discards the partially cached file since partially cached files are not useful (it has to start iterating from the start on the server next time anyway).

Partial caching is useful for development or smaller datasets; it is provided by the .dbcache(fname, size) method or the dbcache.DBCache IterableDataset.

So, in short, use dbcache for development and shard caching for production.

Please reopen the bug if you find a situation where fully completed shards are not cached.

**Comment by kdonbekci** [comment]
This issue happens when fully iterating over the dataset and the shards. The colab notebook demonstrates this issue with fully completed shards, so I am reopening this issue.
